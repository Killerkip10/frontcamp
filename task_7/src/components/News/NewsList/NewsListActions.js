export const NEWS_LIST_GET_REQUEST = 'NEWS_LIST_GET_REQUEST';
export const NEWS_LIST_GET_SUCCESS = 'NEWS_LIST_GET_SUCCESS';
export const NEWS_LIST_GET_FAILURE = 'NEWS_LIST_GET_FAILURE';

export const CHANGE_TOPIC = 'CHANGE_TOPIC';

export const getNewsListRequest = () => ({
  type: NEWS_LIST_GET_REQUEST,
  payload: { isFetching: true },
});

export const getNewsListSuccess = news => ({
  type: NEWS_LIST_GET_SUCCESS,
  payload: { news, isFetching: false },
});

export const getNewsListFailure = () => ({
  type: NEWS_LIST_GET_FAILURE,
  payload: { isFetching: false },
});

export const changeTopic = topic => ({
  type: CHANGE_TOPIC,
  payload: { topic },
});