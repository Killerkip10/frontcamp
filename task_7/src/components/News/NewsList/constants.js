export const ARTICLE_TOPICS = {
  ARTS: 'arts',
  AUTOMOBILES: 'automobiles',
  BOOKS: 'books',
  BUSINESS: 'business',
  FASHION: 'fashion',
  FOOD: 'food',
  HEALTH: 'health',
  HOME: 'home',
  INSIDER: 'insider',
  MAGAZINE: 'magazine',
  MOVIES: 'movies',
  NATIONAL: 'national',
};

export const DATA_CY = {
  LOADING: 'loading',
  SELECT_TOPIC: 'selectTopic',
  LIST_ITEM: 'listItem',
};