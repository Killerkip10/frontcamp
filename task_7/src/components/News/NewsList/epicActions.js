import { API, API_KEY } from 'configs/api';

import {
  getNewsListRequest,
  getNewsListSuccess,
  getNewsListFailure,
} from './actions';

export const getNewsList = (topic) => (dispatch) => {
  dispatch(getNewsListRequest());

  return fetch(`${API.HOST}${API.GET_ARTICLES}${topic}.json?api-key=${API_KEY}`)
    .then(response => response.json())
    .then(({ results }) => dispatch(getNewsListSuccess(results)))
    .catch(() => dispatch(getNewsListFailure()));
};