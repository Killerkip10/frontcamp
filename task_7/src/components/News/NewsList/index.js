export * from './NewsListContainer';
export * from './NewsListReducer';