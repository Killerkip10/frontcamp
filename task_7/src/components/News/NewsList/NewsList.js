import React, { useLayoutEffect, useCallback, memo } from 'react';

import { ARTICLE_TOPICS } from './constants';

export const NewsList = memo(({
  isFetching,
  news,
  topic,
  getNews,
  changeTopic,
  clickDetails,
}) => {
  useLayoutEffect(() => { getNews(topic); }, [topic, getNews]);

  const handleTopicChange = useCallback(({ target: { value } }) => changeTopic(value), [changeTopic]);

  const handleDetailsClick = useCallback(({ target: { dataset: { index } } }) => index && clickDetails(index, topic), [topic]);

  if (isFetching) {
    return <div>...Loading</div>;
  }

  return (
    <div>
      <div>
        <select value={topic} onChange={handleTopicChange}>
          {Object.values(ARTICLE_TOPICS).map(value => (
            <option key={value} value={value}>{value}</option>         
          ))}
        </select>
      </div>

      <div>
        {news.map(({ title, url }, index) => (
          <div key={url} onClick={handleDetailsClick}>
            <div>{title}</div>
            <button data-index={index}>Details</button>
          </div>
        ))}
      </div>
    </div>
  );
});
