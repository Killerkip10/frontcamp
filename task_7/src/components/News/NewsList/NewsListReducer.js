import { ARTICLE_TOPICS } from './constants';
import {
  NEWS_LIST_GET_REQUEST,
  NEWS_LIST_GET_SUCCESS,
  NEWS_LIST_GET_FAILURE,
  CHANGE_TOPIC,
} from './NewsListActions';

export const initialState = {
  isFetching: false,
  topic: ARTICLE_TOPICS.ARTS,
  news: [],
};

export const newsList = (state = initialState, { type, payload }) => {
  switch (type) {
    case NEWS_LIST_GET_REQUEST:
    case NEWS_LIST_GET_SUCCESS:
    case NEWS_LIST_GET_FAILURE:
    case CHANGE_TOPIC:
      return { ...state,  ...payload };
    default:
      return state;
  }
};