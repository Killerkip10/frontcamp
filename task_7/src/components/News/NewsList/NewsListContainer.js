import React, { useCallback, useEffect } from 'react';
import { connect } from 'react-redux';

import { NewsList } from './NewsList';
import { getNewsList } from './NewsListEpicActions';
import { changeTopic } from './NewsListActions';

export const NewsListContainerComponent = ({ history, isFetching, news, topic, getNews, changeTopic }) => {
  useEffect(() => { getNews(topic); }, [topic]);

  const handleDetailsClick = useCallback((id, topic) => history.push(`/news/${id}/${topic}`), [history]);
  
  if (isFetching) {
    return <div>...Loading</div>;
  }

  return (
    <NewsList
      news={news}
      topic={topic}
      changeTopic={changeTopic}
      clickDetails={handleDetailsClick}
    />
  );
};

const mapStateToProps = state => ({
  isFetching: state.news.newsList.isFetching,
  news: state.news.newsList.news,
  topic: state.news.newsList.topic,
});

const mapStateToDispatch = dispatch => ({
  getNews: topic => dispatch(getNewsList(topic)),
  changeTopic: topic => dispatch(changeTopic(topic)),
});

const connector = connect(mapStateToProps, mapStateToDispatch);

export const NewsListContainer = connector(NewsListContainerComponent);