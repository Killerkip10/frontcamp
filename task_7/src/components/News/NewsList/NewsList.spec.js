import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

import { NewsList } from './NewsList';

describe('NewsList Component', () => {
    let props;

    beforeEach(() => {
        props = {
            news: [{
                title: 'title-1',
                url: 'url-1',
            }],
            topic: 'topic-1',
            changeTopic: jest.fn(),
            clickDetails: jest.fn(),
        };
    });

    it('should render NewsList', () => {
        const component = shallow(<NewsList {...props} />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('should call handleTopicChange', () => {
        const value = 'option-1';
        const component = shallow(<NewsList {...props} />);

        component.find('select').simulate('change', { target: { value } });

        expect(props.changeTopic).toBeCalledWith(value);
    });
});