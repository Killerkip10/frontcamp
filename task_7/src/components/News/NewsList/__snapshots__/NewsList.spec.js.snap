// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsList Component should render NewsList 1`] = `
<div>
  <div>
    <select
      data-cy="selectTopic"
      onChange={[Function]}
      value="topic-1"
    >
      <option
        key="arts"
        value="arts"
      >
        arts
      </option>
      <option
        key="automobiles"
        value="automobiles"
      >
        automobiles
      </option>
      <option
        key="books"
        value="books"
      >
        books
      </option>
      <option
        key="business"
        value="business"
      >
        business
      </option>
      <option
        key="fashion"
        value="fashion"
      >
        fashion
      </option>
      <option
        key="food"
        value="food"
      >
        food
      </option>
      <option
        key="health"
        value="health"
      >
        health
      </option>
      <option
        key="home"
        value="home"
      >
        home
      </option>
      <option
        key="insider"
        value="insider"
      >
        insider
      </option>
      <option
        key="magazine"
        value="magazine"
      >
        magazine
      </option>
      <option
        key="movies"
        value="movies"
      >
        movies
      </option>
      <option
        key="national"
        value="national"
      >
        national
      </option>
    </select>
  </div>
  <div>
    <div
      data-cy="listItem"
      key="url-1"
      onClick={[Function]}
    >
      <div>
        title-1
      </div>
      <button
        data-index={0}
      >
        Details
      </button>
    </div>
  </div>
</div>
`;
